File: packages/java/src/main/java/io/axene/mailer/Templates.java
package io.axene.mailer;

import java.util.List;

/** Manage reusable email templates (Starter plan and up). Accessed as {@code client.templates()}. */
public final class Templates {

    private final ApiTransport transport;

    Templates(ApiTransport transport) {
        this.transport = transport;
    }

    /**
     * List all templates, most recently updated first.
     *
     * @return the templates.
     */
    public List<Template> list() {
        return transport.request("GET", "/v1/templates/", null, transport.listType(Template.class));
    }

    /**
     * Create a template. {@code variables} are derived server-side from
     * {@code {{name}}} placeholders, so you do not pass them.
     *
     * @param params the template fields ({@code name} required).
     * @return the created template.
     */
    public Template create(TemplateParams params) {
        return transport.request("POST", "/v1/templates/", params.toWire(), transport.type(Template.class));
    }

    /**
     * Fetch a single template.
     *
     * @param id the template id.
     * @return the template.
     */
    public Template get(String id) {
        return transport.request("GET", "/v1/templates/" + Query.enc(id), null, transport.type(Template.class));
    }

    /**
     * Update a template (partial).
     *
     * @param id     the template id.
     * @param params the fields to update.
     * @return the updated template.
     */
    public Template update(String id, TemplateParams params) {
        return transport.request("PATCH", "/v1/templates/" + Query.enc(id), params.toWire(), transport.type(Template.class));
    }

    /**
     * Delete a template.
     *
     * @param id the template id.
     */
    public void delete(String id) {
        transport.request("DELETE", "/v1/templates/" + Query.enc(id), null, transport.type(Void.class));
    }

    /**
     * Duplicate a template (the copy's {@code blocks_json} is not carried over).
     *
     * @param id the template id.
     * @return the duplicated template.
     */
    public Template duplicate(String id) {
        return transport.request("POST", "/v1/templates/" + Query.enc(id) + "/duplicate", null, transport.type(Template.class));
    }
}

File: packages/php/src/Resources/Templates.php
<?php

declare(strict_types=1);

namespace Axene\Mailer\Resources;

use Axene\Mailer\Internal;
use Axene\Mailer\Transport;

/**
 * The `templates` resource: reusable email templates (Starter plan and up).
 * Accessed as `$client->templates`.
 *
 * The SDK's `html` and `text` map to the wire fields `html_body` and
 * `text_body` (templates only; emails keep `html`/`text`).
 */
final class Templates
{
    /** @internal */
    public function __construct(private readonly Transport $transport)
    {
    }

    /**
     * List all templates, most recently updated first.
     *
     * @return list<array<string, mixed>>
     */
    public function list(): array
    {
        return $this->transport->request('GET', '/v1/templates/');
    }

    /**
     * Create a template. `variables` are derived server-side from `{{name}}`
     * placeholders in the bodies, so you do not pass them.
     *
     * @param array{name: string, subject?: string|null, html?: string|null, text?: string|null, blocksJson?: array<string, mixed>|null} $params
     * @return array<string, mixed>
     */
    public function create(array $params): array
    {
        return $this->transport->request('POST', '/v1/templates/', Internal::prune([
            'name' => $params['name'] ?? null,
            'subject' => $params['subject'] ?? null,
            'html_body' => $params['html'] ?? null,
            'text_body' => $params['text'] ?? null,
            'blocks_json' => $params['blocksJson'] ?? null,
        ]));
    }

    /**
     * Fetch a single template.
     *
     * @return array<string, mixed>
     */
    public function get(string $id): array
    {
        return $this->transport->request('GET', '/v1/templates/' . rawurlencode($id));
    }

    /**
     * Update a template (partial).
     *
     * @param array{name?: string, subject?: string|null, html?: string|null, text?: string|null, blocksJson?: array<string, mixed>|null} $params
     * @return array<string, mixed>
     */
    public function update(string $id, array $params): array
    {
        return $this->transport->request('PATCH', '/v1/templates/' . rawurlencode($id), Internal::prune([
            'name' => $params['name'] ?? null,
            'subject' => $params['subject'] ?? null,
            'html_body' => $params['html'] ?? null,
            'text_body' => $params['text'] ?? null,
            'blocks_json' => $params['blocksJson'] ?? null,
        ]));
    }

    /**
     * Delete a template.
     */
    public function delete(string $id): void
    {
        $this->transport->request('DELETE', '/v1/templates/' . rawurlencode($id));
    }

    /**
     * Duplicate a template (the copy's `blocks_json` is not carried over).
     *
     * @return array<string, mixed>
     */
    public function duplicate(string $id): array
    {
        return $this->transport->request('POST', '/v1/templates/' . rawurlencode($id) . '/duplicate');
    }
}
